scifight/utils.py

```diff
@@ -0,0 +1,34 @@
+
+def shorten_text(text, maxchars, ellipsis='...'):
+    """ Returns first few words from the `text`, but no more than `maxchars`, and appends
+        an `ellipsis`. This function works only on plain text (no HTML!), removes any leading
+        and trailing spaces, and squashes any number of consequent whitespaces (including
+        newlines) into a single space.
+
+        :param text: Input string, possibly quite long.
+        :param maxchars: Maximum length or result string.
+        :param ellipsis: Trailing sign to indicate implied continuation.
+        :return: Shortened substring with squashed whitespaces.
+    """
+    text = text.strip().replace('\n', ' ')
+    new_text = ''
+    have_space = False
+    for char in text:
+        if char == ' ':
+            if not have_space:
+                new_text += char
```

Comment:
Potentially intensive memory reallocations at this point. You better just count the number of characters instead immediate appending them.